File: tools/bill_pipe.py
"""Independent planar reference for Bill's original metal connecting pipe.

The octagonal surface uses no additional voxel skin. Its two collars derive
from the endpoint artwork; the long center strip stays a smooth tube.
"""
from math import floor, sqrt


def build(sp, t):
    variant = t["pipe_variant"]
    assert variant in ("link", "left", "right")
    assert (sp["W"], sp["H"]) == ((48 if variant == "link" else 8), 16)
    if variant == "link":
        grey, white, dark, black = (8, 6), (8, 7), (8, 10), (8, 5)
    else:
        grey, white, dark, black = (4, 3), (4, 6), (4, 10), (4, 2)
    for donor, color in zip((grey, white, dark, black), (1, 0, 2, 3)):
        x, y = donor
        assert sp["col"][y][x] == color and not sp["out"][y][x]

    # Interpret the nine-pixel tube and eleven-pixel cuff elevations as
    # actual octagonal cross-sections, with their original8.5px center.
    # Its depth axis aligns with the accepted chamber center at world32.
    k = sqrt(2) - 1
    ring = ((1, k), (k, 1), (-k, 1), (-1, k),
            (-1, -k), (-k, -1), (k, -1), (1, -k))
    materials = (white, white, grey, dark, dark, grey, grey, grey)
    xmin, xmax, collars = {
        "link": (-4, 52, ((4, 7), (41, 44))),
        "left": (0, 12, ((1, 4),)),
        "right": (-4, 8, ((4, 7),)),
    }[variant]
    stations = {xmin, xmax}
    for a, b in collars:
        stations.update((a, a + .35, b - .35, b))
    # Rendering uses the shared8px world-curvature lattice. Intermediate
    # stations add no collars and do not alter the source-metal pattern.
    stations.update(range((floor(xmin / 8) + 1) * 8, xmax, 8))
    stations = sorted(stations)
    sections = []
    for a, b in zip(stations, stations[1:]):
        mid = (a + b) / 2
        cuff = next(((lo, hi) for lo, hi in collars if lo < mid < hi), None)
        radius = 5.5 if cuff else 4.5
        band = bool(cuff and (mid < cuff[0] + .35 or mid > cuff[1] - .35))
        sections.append((a, b, radius, band))

    def point(x, r, j):
        dy, dz = ring[j % 8]
        return (x, 8.5 + r * dy, 16 + r * dz)

    faces = []

    def face(points, source, shade):
        faces.append(dict(points=points, source=source, shade=shade))

    previous = None
    for a, b, radius, band in sections:
        if previous is not None and previous != radius:
            for j in range(8):
                face([point(a, previous, j), point(a, previous, j + 1),
                      point(a, radius, j + 1), point(a, radius, j)], black, .75)
        for j in range(8):
            ny = ring[j][0] + ring[(j + 1) % 8][0]
            nz = ring[j][1] + ring[(j + 1) % 8][1]
            length = sqrt(ny * ny + nz * nz)
            shade = .68 + .27 * max(0, ny / length) + .1 * max(0, nz / length)
            face([point(a, radius, j), point(a, radius, j + 1),
                  point(b, radius, j + 1), point(b, radius, j)],
                 black if band else materials[j], shade)
        previous = radius

    # Four ordinary quads close each end; all vertices are distinct.
    for left, x in ((True, xmin), (False, xmax)):
        for j in range(0, 8, 2):
            vertices = [point(x, 4.5, n) for n in range(j, j + 3)]
            if left:
                vertices.reverse()
            face([(x, 8.5, 16)] + vertices, dark if left else grey,
                 .68 if left else .78)
    return faces

File: lib/BillPipe.lua
-- Bill's exposed silver pipe: a low-cost octagonal shell with the source
-- drawing's two raised joint collars. The former wall boxes are claimed
-- separately; only original metal texels are sampled here.
local Pipe = {}
function Pipe.model(sp, pr, t)
  local kind = t.pipeVariant
  assert(kind == "link" or kind == "left" or kind == "right", "Bill pipe variant")
  assert(sp.H == 16 and sp.W == (kind == "link" and 48 or 8), "Bill pipe source dimensions")
  local faces, W = {}, sp.W
  local function source(x,y)
    local i=y*W+x
    assert(sp.inside[i], "pipe material belongs to its original metal silhouette")
    return i
  end
  local grey,white,dark,black
  if kind=="link" then
    grey,white,dark,black=source(8,6),source(8,7),source(8,10),source(8,5)
  else
    grey,white,dark,black=source(4,3),source(4,6),source(4,10),source(4,2)
  end
  assert(sp.col[grey]==1 and sp.col[white]==0 and sp.col[dark]==2 and sp.col[black]==3)
  -- The source tube occupies elevation rows3..11, with its collar in
  -- rows2..12: diameter9 and11, centered8.5 above ground. Put its axis at
  -- the chambers' real center depth (world32 from this pair's origin16).
  -- The4px inserts meet the narrower cylindrical body under its cap.
  local cy,cz=8.5,16
  local oct={{1,math.sqrt(2)-1},{math.sqrt(2)-1,1},{1-math.sqrt(2),1},
    {-1,math.sqrt(2)-1},{-1,1-math.sqrt(2)},{1-math.sqrt(2),-1},
    {math.sqrt(2)-1,-1},{1,1-math.sqrt(2)}}
  local function p(x,r,j)
    local v=oct[(j-1)%8+1]
    return{x,cy+r*v[1],cz+r*v[2]}
  end
  local function face(a,b,c,d,i,shade)
    faces[#faces+1]={a,b,c,d,source=i,shade=shade}
  end
  local materials={white,white,grey,dark,dark,grey,grey,grey}
  local function shell(a,b,r,band)
    for j=1,8 do
      local n=oct[j];local next=oct[j%8+1]
      local ny,nz=n[1]+next[1],n[2]+next[2]
      local length=math.sqrt(ny*ny+nz*nz);ny,nz=ny/length,nz/length
      local shade=0.68+0.27*math.max(0,ny)+0.1*math.max(0,nz)
      face(p(a,r,j),p(a,r,j+1),p(b,r,j+1),p(b,r,j),band and black or materials[j],shade)
    end
  end
  local function shoulder(x,old,new)
    for j=1,8 do
      face(p(x,old,j),p(x,old,j+1),p(x,new,j+1),p(x,new,j),black,0.75)
    end
  end
  local sections={}
  local function section(a,b,r,band)sections[#sections+1]={a,b,r,band}end
  local function collar(a,b)
    section(a,a+0.35,5.5,true)
    section(a+0.35,b-0.35,5.5)
    section(b-0.35,b,5.5,true)
  end
  if kind=="link" then
    section(-4,4,4.5);collar(4,7);section(7,41,4.5);collar(41,44);section(44,52,4.5)
  elseif kind=="left" then
    section(0,1,4.5);collar(1,4);section(4,12,4.5)
  else
    section(-4,4,4.5);collar(4,7);section(7,8,4.5)
  end
  local last
  for _,s in ipairs(sections)do
    if last and last[3]~=s[3]then shoulder(s[1],last[3],s[3])end
    -- Buildings bends the world on an8px lattice. Keep this smooth tube
    -- on the same axial segmentation, without adding visible joint ribs.
    local x=s[1]
    while x<s[2]do
      local next=math.min(s[2],(math.floor(x/8)+1)*8)
      shell(x,next,s[3],s[4]);x=next
    end
    last=s
  end
  -- Four planar quads close each octagonal end without degenerate
  -- triangles or redundant internal caps between tube/collar sections.
  for n,s in ipairs({sections[1],sections[#sections]})do
    local x=s[n==1 and 1 or 2];local center={x,cy,cz}
    for j=1,8,2 do
      if n==1 then face(center,p(x,s[3],j+2),p(x,s[3],j+1),p(x,s[3],j),dark,0.68)
      else face(center,p(x,s[3],j),p(x,s[3],j+1),p(x,s[3],j+2),grey,0.78)end
    end
  end
  -- No proxy is needed for this isolated closed shell. Empty occupancy
  -- avoids emitting a second voxel skin; the standard surface path uses
  -- the original atlas and shared shader for all faces.
  return{W=W,ytop=13,zmin=10,zmax=21,at=function()return nil end,surfaces=faces}
end
return Pipe
